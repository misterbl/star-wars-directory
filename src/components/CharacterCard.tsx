import React, { PureComponent } from "react";
import charactersPhotos from "../const/charactersPhotos";
import { withRouter, RouteComponentProps, StaticContext } from "react-router";
import { bindActionCreators, Dispatch } from "redux";
import { connect } from "react-redux";
import { ICharacterCardComponent, ICharacterCard } from "./CharacterCard.d";
import ROUTES from "../const/routes";
import getCharacterDetails from "../actions/thunks/getCharacterDetails";
import { assignCurrentView } from "../actions/actionCreators/actions";

export class CharacterCard extends PureComponent<ICharacterCardComponent> {
  onClick = async () => {
    const {
      getCharacterDetails,
      assignCurrentView,
      character,
      history: { push }
    } = this.props;
    const { species, films, vehicles, homeworld, starships } = character;
    await assignCurrentView(character);

    await getCharacterDetails(
      species[0],
      homeworld,
      films,
      vehicles,
      starships
    );

    push(ROUTES.CHARACTER);
  };
  render() {
    const { character } = this.props;
    const { name, birth_year, gender, height } = character;
    const sprite = charactersPhotos[character.name];
    return (
      <div
        role="button"
        onClick={this.onClick}
        className="pointer m-3 text-white border border-dark p-3"
      >
        <u className="card-text">{name.toUpperCase()}</u>
        <div className="d-flex">
          {sprite ? (
            <div className="avatar my-2">
              <div
                style={{
                  backgroundPosition: `${sprite}`
                }}
              />
            </div>
          ) : (
            <div className="avatar avatar__default my-2" />
          )}
          <div className="ml-3">
            <p className="m-1 text-secondary">{`Born in ${birth_year}`}</p>
            <p className="m-1 text-secondary">{`Gender: ${gender}`}</p>
            <p className="m-1 text-secondary">{`Height: ${height}cm`}</p>
          </div>
        </div>
      </div>
    );
  }
}

export const mapDispatchToProps = (dispatch: Dispatch) => ({
  getCharacterDetails: bindActionCreators(getCharacterDetails, dispatch),
  assignCurrentView: bindActionCreators(assignCurrentView, dispatch)
});

export default withRouter<
  ICharacterCard & RouteComponentProps<any, StaticContext>
>(
  connect(
    null,
    mapDispatchToProps
  )(CharacterCard)
);
