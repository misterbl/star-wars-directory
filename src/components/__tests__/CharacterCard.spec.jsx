/* eslint-disable no-undef */
import * as React from "react";
import { shallow } from "enzyme";
import history from "../../testMocks/history.mock";
import { CharacterCard } from "../CharacterCard";
import ROUTES from "../../routes";

describe("CharacterCard", () => {
  const props = {
    history,
    singleSearch: jest.fn(),
    character: {
      name: "name",
      height: "123",
      mass: "12",
      birth_year: "1234",
      gender: "gender"
    }
  };
  const wrapper = shallow(<CharacterCard {...props} />);
  it("matches the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("calls singleSearch() onClick ", () => {
    const mainDiv = wrapper.find("div").at(0);
    mainDiv.simulate("click");
    expect(props.singleSearch).toHaveBeenCalled();
  });
  it("calls push() onClick ", () => {
    const mainDiv = wrapper.find("div").at(0);
    mainDiv.simulate("click");
    expect(props.history.push).toHaveBeenCalledWith(ROUTES.CHARACTER);
  });
});
