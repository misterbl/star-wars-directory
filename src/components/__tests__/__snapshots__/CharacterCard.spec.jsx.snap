// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCard matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CharacterCard
    character={
      Object {
        "birth_year": "1234",
        "gender": "gender",
        "height": "123",
        "mass": "12",
        "name": "name",
      }
    }
    history={
      Object {
        "action": "PUSH",
        "block": [MockFunction],
        "createHref": [MockFunction],
        "go": [MockFunction],
        "goBack": [MockFunction],
        "goForward": [MockFunction],
        "length": 123,
        "listen": [MockFunction],
        "location": Object {
          "hash": "",
          "key": "",
          "pathname": "",
          "search": "",
          "state": "",
        },
        "push": [MockFunction],
        "replace": [MockFunction],
      }
    }
    singleSearch={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "name",
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <u
          className="card-text"
        >
          NAME
        </u>,
        <div
          className="d-flex"
        >
          <div
            className="avatar avatar__default card-img-top"
          />
          <div
            className="ml-3"
          >
            <p
              className="m-1 text-secondary"
            >
              Born in 1234
            </p>
            <p
              className="m-1 text-secondary"
            >
              Gender: gender
            </p>
            <p
              className="m-1 text-secondary"
            >
              Height: 123cm
            </p>
          </div>
        </div>,
      ],
      "className": "pointer m-3 text-white",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "NAME",
          "className": "card-text",
        },
        "ref": null,
        "rendered": "NAME",
        "type": "u",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="avatar avatar__default card-img-top"
            />,
            <div
              className="ml-3"
            >
              <p
                className="m-1 text-secondary"
              >
                Born in 1234
              </p>
              <p
                className="m-1 text-secondary"
              >
                Gender: gender
              </p>
              <p
                className="m-1 text-secondary"
              >
                Height: 123cm
              </p>
            </div>,
          ],
          "className": "d-flex",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "avatar avatar__default card-img-top",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <p
                  className="m-1 text-secondary"
                >
                  Born in 1234
                </p>,
                <p
                  className="m-1 text-secondary"
                >
                  Gender: gender
                </p>,
                <p
                  className="m-1 text-secondary"
                >
                  Height: 123cm
                </p>,
              ],
              "className": "ml-3",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Born in 1234",
                  "className": "m-1 text-secondary",
                },
                "ref": null,
                "rendered": "Born in 1234",
                "type": "p",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Gender: gender",
                  "className": "m-1 text-secondary",
                },
                "ref": null,
                "rendered": "Gender: gender",
                "type": "p",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Height: 123cm",
                  "className": "m-1 text-secondary",
                },
                "ref": null,
                "rendered": "Height: 123cm",
                "type": "p",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "name",
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <u
            className="card-text"
          >
            NAME
          </u>,
          <div
            className="d-flex"
          >
            <div
              className="avatar avatar__default card-img-top"
            />
            <div
              className="ml-3"
            >
              <p
                className="m-1 text-secondary"
              >
                Born in 1234
              </p>
              <p
                className="m-1 text-secondary"
              >
                Gender: gender
              </p>
              <p
                className="m-1 text-secondary"
              >
                Height: 123cm
              </p>
            </div>
          </div>,
        ],
        "className": "pointer m-3 text-white",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "NAME",
            "className": "card-text",
          },
          "ref": null,
          "rendered": "NAME",
          "type": "u",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="avatar avatar__default card-img-top"
              />,
              <div
                className="ml-3"
              >
                <p
                  className="m-1 text-secondary"
                >
                  Born in 1234
                </p>
                <p
                  className="m-1 text-secondary"
                >
                  Gender: gender
                </p>
                <p
                  className="m-1 text-secondary"
                >
                  Height: 123cm
                </p>
              </div>,
            ],
            "className": "d-flex",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "avatar avatar__default card-img-top",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <p
                    className="m-1 text-secondary"
                  >
                    Born in 1234
                  </p>,
                  <p
                    className="m-1 text-secondary"
                  >
                    Gender: gender
                  </p>,
                  <p
                    className="m-1 text-secondary"
                  >
                    Height: 123cm
                  </p>,
                ],
                "className": "ml-3",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Born in 1234",
                    "className": "m-1 text-secondary",
                  },
                  "ref": null,
                  "rendered": "Born in 1234",
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Gender: gender",
                    "className": "m-1 text-secondary",
                  },
                  "ref": null,
                  "rendered": "Gender: gender",
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Height: 123cm",
                    "className": "m-1 text-secondary",
                  },
                  "ref": null,
                  "rendered": "Height: 123cm",
                  "type": "p",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
